### PR TITLE
Tree: Implement moves in sequence field

### PR DESCRIPTION
**Pull request by alex-pardes**

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/sequence-field/utils.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/utils.ts:315]: Is `moveEffects.dstMergeable.get(prevMerge)` expected to always be undefined here? If so, how do we know that is it?

File: packages/dds/tree/src/feature-libraries/sequence-field/utils.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/utils.ts:385]: It would be very helpful to document each member.

File: packages/dds/tree/src/feature-libraries/sequence-field/utils.ts
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/utils.ts:315]: I'm also having trouble figuring out what it means for `prevMerge` to not be undefined. It seems to mean we've tried to merge `lhsMoveIn` with another mark already, but if we did, how come the lhs mark is not that other mark?

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/utils.ts:315]: This is intended to handle the case where we merge more than two adjacent changes. Suppose there are three mergeable moves, A, B, and C. We might first come across the move-outs, where we will record that A could be merged with B, and B with C on the source side, but we can't actually merge the move-outs yet because we don't know if the destination side is mergeable. When we process the move-ins we will merge B into A and then C into A. When we process the move-outs again in the next pass we will need to remember that A, B, and C all could be merged on the destination side. So while handling the destination merges, when we merge C into A we check that A was already merged with B, and mark that B and C are mergeable instead of A and C.

File: packages/dds/tree/src/feature-libraries/sequence-field/invert.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/invert.ts:89]: We also need to inver the `mark.changes` if any.
The idea of having detach marks in detached trees and attach marks in attached trees was in part motivated by the fact that it simplifies invert.

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/invert.ts:89]: Ah, right. My preference is to just to transfer the child marks. It should be very simple compared to the move handling in rebase and compose, and I'd rather have a little extra complexity in `invert` than extra complexity in the contract for the format.

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts:69]: It seems to make the API easier to work with if we do.
Also, for the `sourceIndex === destIndex`, it can have effects in the presence of concurrent changes (it should bring back those nodes that had been concurrently moved away), which makes me think `return []` is not what we want for that case.

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts:86]: This doesn't take into account the possibility that `sourceIndex + count > destIndex`. Maybe add a TODO.

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts:128]: This doesn't take into account the possibility that `sourceIndex + count > destIndex`. Maybe add a TODO.

File: packages/dds/tree/src/feature-libraries/sequence-field/invert.ts
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/invert.ts:89]: Sounds good.

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts:86]: Can you elaborate on what the problem is?

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:265]: This is always `undefined` in the current implementation. Is that by design?

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:265]: We should pass `this.baseRevision` in `getNextBaseMark`.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:373]: Why do we need to do this?

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:373]: Suppose we have the following mark list: `[4, insert, 3, delete(2)]` and we want to move the delete to index 1. The outcome should be `[1, delete(2), 1, insert]`. Nodes 2 and 3 were previously skipped, but now they are deleted. The negative offset is a somewhat kludgy way of removing the skip marks we are replacing with the moved marks.

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldEditor.ts:86]: Ah, sorry, I forgot about the fact that the destIndex is "after removal".
What I should have done is comment that the API prevents creating moves where the destination index lies between the detached nodes.
E.g., in a sequence [A, B, C], move(src: 0, count: 3, dst: 1).
This is a little odd but it is well-formed and meaningfully different from (say) move(src: 0, count: 3, dst: 0).
So we should either support such moves or provide a good reason why we don't want to allow them.

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:74]: We need to deal with changes here if any. I'm guessing this will happen in a separate PR. Feel free to resolve/ignore without further action.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:386]: Yes. I was wondering why this isn't caught by the rebase tests, but I realized it's because this whole block is now redundant. `ComposeQueue.pop` splits the marks before returning them.

File: packages/dds/tree/src/feature-libraries/sequence-field/compose.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/compose.ts:180]: `newMark.revision ?? newRev` is used 11 times in this function. Computing it once after `if (isSkipMark(newMark)) { ... }` would help tidy up.

File: packages/dds/tree/src/feature-libraries/sequence-field/compose.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/compose.ts:398]:
```suggestion
    if (!moveEffects.validatedMarks.has(mark)) {
```

File: packages/dds/tree/src/test/feature-libraries/sequence-field/invert.spec.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/invert.spec.ts:76]: We need a Return => Return test as well

File: packages/dds/tree/src/test/feature-libraries/sequence-field/markListFactory.spec.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/markListFactory.spec.ts:73]: Did you forget to re-enable these tests?

File: packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:371]: If all three operations are concurrent, wouldn't C be at index 1?

File: packages/dds/tree/src/test/feature-libraries/sequence-field/markListFactory.spec.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/markListFactory.spec.ts:91]:
```suggestion
        assert.deepEqual(factory2.list, [
```

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/feature-libraries/sequence-field/markListFactory.spec.ts:91]: Ah, I see this whole file needs to be updated in this way. Feel free to ignore.

File: packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts
Comment (alex-pardes) [reply packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:371]: I don't understand the question, but C is moving a node to index 1.

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:371]: I was assuming a starting state [ACD]. I'm guessing it's [ADC] instead.
It seems like the end result is [BCD], shouldn't it be [CBD] instead since the move-in is sequenced after the insert and everything has merge-left policy?

**Review by yann-achard-MS — APPROVED** [review]
Approving so we can merge this in before the holiday break.
The remaining comments and other cleanup will follow later.

File: packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts
Comment (alex-pardes) [reply packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:371]: I don't follow your notation, but the intention for this test was to check that we are able to retain the knowledge that the move-in goes to the right of the insert (since they are in the same position after rebasing over the delete).

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:371]: Oh I get it. For what it's worth, I've been using "A", "B", "C", etc. to refer to nodes in the document, where the nodes are expected to end up in alphabetical order. I got confused about the test because I trusted that convention when I shouldn't have. It seems like you're just using the letters to refer to the changes instead.
Anyway, mystery solved. :)

File: packages/dds/tree/src/feature-libraries/sequence-field/utils.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/utils.ts:767]: The operator === has higher precendence than operator ??, so `mark.revision ?? revision === undefined` is equivalent to `mark.revision ?? (revision === undefined)`, which is probably not want you wanted here.

File: packages/dds/tree/src/feature-libraries/sequence-field/compose.ts
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/compose.ts:403]: I think we also want to check that `cloned.revision === undefined` otherwise we'll end up overwritting it.

**Pull request merged**
